frame/dvm/evm/src/runner/stack.rs

```diff
@@ -66,50 +67,62 @@ impl<T: Config> Runner<T> {
 		) -> (ExitReason, R),
 	{
 		let base_fee = T::FeeCalculator::min_gas_price();
-		// Gas price check is skipped when performing a gas estimation.
-		let max_fee_per_gas = match max_fee_per_gas {
-			Some(max_fee_per_gas) => {
-				ensure!(max_fee_per_gas >= base_fee, <Error<T>>::GasPriceTooLow);
+		let max_fee_per_gas = match (max_fee_per_gas, is_transactional) {
+			(Some(max_fee_per_gas), _) => {
+				ensure!(max_fee_per_gas >= base_fee, Error::<T>::GasPriceTooLow);
 				max_fee_per_gas
 			},
-			None => Default::default(),
+			// For the internal transaction, the gas_price is None and is_transactional is true
+			// which is legal in our chain. Note that non-internal transaction with gas_price None
+			// will be rejected before entering the runner.
+			// TODO: Delete this after s2s related pallets are removed later.
+			(None, true) => Default::default(),
+			// Gas price check is skipped for non-transactional calls that don't
+			// define a `max_fee_per_gas` input.
+			(None, false) => Default::default(),
 		};
 
-		let vicinity = Vicinity { gas_price: max_fee_per_gas, origin: source };
-
-		let metadata = StackSubstateMetadata::new(gas_limit, &config);
-		let state = SubstrateStackState::new(&vicinity, metadata);
-		let mut executor = StackExecutor::new_with_precompiles(state, config, precompiles);
+		if let Some(max_priority_fee) = max_priority_fee_per_gas {
+			ensure!(max_fee_per_gas >= max_priority_fee, Error::<T>::GasPriceTooLow);
+		}
 
 		// After eip-1559 we make sure the account can pay both the evm execution and priority fees.
-		let max_base_fee =
-			max_fee_per_gas.checked_mul(U256::from(gas_limit)).ok_or(ArithmeticError::Overflow)?;
-		let max_priority_fee = if let Some(max_priority_fee) = max_priority_fee_per_gas {
-			max_priority_fee.checked_mul(U256::from(gas_limit)).ok_or(<Error<T>>::FeeOverflow)?
-		} else {
-			U256::zero()
-		};
 		let total_fee =
-			max_base_fee.checked_add(max_priority_fee).ok_or(<Error<T>>::FeeOverflow)?;
+			max_fee_per_gas.checked_mul(U256::from(gas_limit)).ok_or(Error::<T>::FeeOverflow)?;
 
 		let total_payment = value.checked_add(total_fee).ok_or(Error::<T>::PaymentOverflow)?;
 		let source_account = Pallet::<T>::account_basic(&source);
-		ensure!(source_account.balance >= total_payment, <Error<T>>::BalanceLow);
+		// Account balance check is skipped if fee is Zero.
+		// This case is previously verified to only happen on either:
+		// 	- Non-transactional calls.
+		// 	- BaseFee is configured to be Zero.
+		if total_fee > U256::zero() {
+			ensure!(source_account.balance >= total_payment, Error::<T>::BalanceLow);
+		}
```

Comment:
Should we add a `TODO` here, supposed to be removed after `internal_transact` be removed.

BTW, why we need this check? It looks dangerous cause possibilities of bypass by accident.

Reply:
For example, some internal transact with non-zero `msg.value`, but `source_account.balance` is zero?

Reply:
> Should we add a TODO here, supposed to be removed after internal_transact be removed.

The validation logic here is consistent with that upstream as much as possible. 

> BTW, why we need this check? It looks dangerous cause possibilities of bypass by accident.

The upstream evm pallet is designed to be integrated independently into the runtime. Obviously, there is some overlap between the checks here and in the ethereum pallet, but these checks would be required if there were no ethereum pallet.

> For example, some internal transact with non-zero msg.value, but source_account.balance is zero?

The internal transaction only accepts input filed from outside, and the value is always zero now. https://github.com/darwinia-network/darwinia-common/blob/main/frame/dvm/ethereum/src/lib.rs#L1066

Reply:
Could you provide the link to the upstream implementation?

Reply:
I mean this check `if total_fee > U256::zero()`

Reply:
https://github.com/darwinia-network/frontier/blob/darwinia-v0.12.4/frame/evm/src/runner/stack.rs#L100-L105

Reply:
OK, but the upstream check `if total_fee > U256::zero()` looks not good and not necessary to me.